README.md
This is a fork of the original [yaze 1.14][yaze]. There are no changes
or fixes here. It was place here to make easy to browse. Also this
emulator has a very useful test suite comparing the results against
a real chip.

[yaze]: ftp://ftp.ping.de/pub/misc/emulators/

Yet Another Z80 Emulator
========================

Yaze is a Z80 and CP/M emulator designed to run on Unix systems.

The package consists of:

* an instruction set simulator which accurately imitates a real Z80
  microprocessor,

* a CP/M-2.2 bios written in C which runs on the Unix host but
  interacts with the simulated Z80,

* a monitor which loads CP/M into the simulated processor's ram and
  which makes Unix directories or files look like CP/M disks, and

* a separate program (cdm) which creates and manipulates CP/M disk
  images for use with yaze.

Yaze's "selling points", compared with previously available Z80
emulators, are:

* It emulates all documented and most undocumented processor-internal
  instructions and all flag bits (all 8, including the undocumented
  ones).  A test program is included in the package which exhaustively
  (well, nearly exhaustively) compares all instructions, for all
  machine states before and after execution, against the results from
  a real chip.

* It is independent of the host machine architecture and instruction
  set.  Written in ANSI standard C it is provided with full source
  code under the GNU General Public License.

* It supports all CP/M disk geometries in the form of images in Unix
  files or as read-only disks constructed on-the-fly from the contents
  of Unix directories.  These disks are indistinguishable from real
  disks for even the most inquisitive, low-level CP/M programs.  They
  can be mounted and unmounted at will during emulation.



Frank Cringle, October 1995
fdc@cliwe.ping.de


Various trademarks are the property of various organisations.


ChangeLog
**** Version 1.14 released ****

2004-04-23  Frank D. Cringle  <fdc@sabre.ping.de>

	* yaze.c, Makefile (VERSION): update version string.

	* simz80.pl: simulate the interrupt register (IR).  (Michael Haardt).

	* bios.c (bios): case 254 incremented pc twice in a macro argument
	(Carl Mascott).

**** Version 1.13 released ****

2004-01-24  Frank D. Cringle  <fdc@cliwe.ping.de>

	* yaze.c, Makefile (VERSION): update version string.

	* simz80.h: pagetable definition did not match the definition in
	yaze.c (Michael Haardt).

**** Version 1.12 released ****

2004-01-12  Frank D. Cringle  <fdc@cliwe.ping.de>

	* Makefile: host2cpm.tcl should be included in the tarball.

**** Version 1.11 released ****

2004-01-11  Frank D. Cringle  <fdc@cliwe.ping.de>

	* monitor.c (CLK_TCK): conditional definition.

	* cdm.c: add 'ambiguous filename -> unix directory' functionality,
	contributed by Carl Mascott.

	* bios.c bios.h monitor.c: apply patch from Guenter Radestock so
	that stdin can be a pipe.
	(readsec): fix lru sort bug reported by Carl Mascott.

	* simz80.pl (OUTP): fix register selection (Michael Haardt).

	* io.c yaze.c: fixes for CP/M 3.0 by Michael Haardt.

	* host2cpm.tcl: copies host directories to/from a cpm disk image
	file, contributed by Gary L. Howell.

	* Makefile: 'make tar' modified so that the test sources are
	converted from LF to CRLF.  Add host2cpm.tcl to DISTRIB.




**** Version 1.10 released ****

1998-01-27  Frank D. Cringle  <fdc@cliwe.ping.de>

	* simz80.pl: set the flags correctly for LD A,I/R and
	CPI/CPIR/CPD/CPDR (reported by Sean Young).

	* monitor.c (mountdir): fix fencepost error when setting cp[15],
	as reported by Richard Hirst.

1998-01-25  Frank D. Cringle  <fdc@cliwe.ping.de>

	* simz80.pl: change direct and indirect references to the RAM()
	macro, because it now evaluates its operand more than once if MMU
	is in use (contributed by Michael Haardt).

	* yaze.c, simz80.h: Make BIOS optional and add MMU initialisation
	(contributed by Michael Haardt).

Sun Apr 28 11:30:52 1996  Frank D. Cringle  <fdc@cliwe.ping.de>

	* cdm.c (getblock): rename index to idx (avoids a warning about
	shadowing index()).
	(main): correct a type in the !USE_GNU_READLINE case.

Thu Apr 25 15:32:24 1996  Frank D. Cringle  <fdc@cliwe.ping.de>

	* bios.c (constat, lststat): select() may modify timeval, so
	reinitialise it.
	(readsec,writesec): protect against undefined or unmounted disk.
	(bios): Add BgiiFix.

	* monitor.c (mount): use MAP_FILE flag in mmap() call (linux-ism?).

	* yaze.c (load_cpm): correctly calculate available memory if
	loadadr is specified.

	* simz80.pl (CALLC): rename symbol adr to avoid warning msg.

	* cdm.c (srchdir): ignore cpm-3 byte count.




bios.c
/* Basic i/o system module.
   Copyright (C) 1995  Frank D. Cringle.

This file is part of yaze - yet another Z80 emulator.

Yaze is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 2 of the License, or (at your
option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <limits.h>
#include <fcntl.h>
#include <termios.h>
#include <ctype.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/stat.h>
#include <dirent.h>
#include <sys/mman.h>

#include "simz80.h"
#include "yaze.h"
#include "bios.h"

/* 	$Id: bios.c,v 1.4 2004/04/23 09:50:15 fdc Exp $	 */

#ifndef lint
static char vcid[] = "$Id: bios.c,v 1.4 2004/04/23 09:50:15 fdc Exp $";
#endif /* lint */


/* Z80 registers */
#define AF	af[af_sel]
#define BC	regs[regs_sel].bc
#define DE	regs[regs_sel].de
#define HL	regs[regs_sel].hl

/* location on current disk */
static WORD	track;
static WORD	sector;
static WORD	dma;

typedef unsigned long COUNT;
static COUNT	bioscount;
static BYTE	old_iobyte;

static void
new_iobyte(int io)
{
    switch ((io >> 6) & 3) {
    case 0:	chn[CHNlst] = TTYout;
	break;
    case 1:	chn[CHNlst] = CRTout;
	break;
    case 2:	chn[CHNlst] = LPTout;
	break;
    case 3:	chn[CHNlst] = UL1out;
	break;
    }
    switch ((io >> 4) & 3) {
    case 0: chn[CHNpun] = TTYout;
	break;
    case 1: chn[CHNpun] = PUNout;
	break;
    case 2: chn[CHNpun] = UP1out;
	break;
    case 3: chn[CHNpun] = UP2out;
	break;
    }
    switch ((io >> 2) & 3) {
    case 0:	chn[CHNrdr] = TTYin;
	break;
    case 1:	chn[CHNrdr] = RDRin;
	break;
    case 2:	chn[CHNrdr] = UR1in;
	break;
    case 3:	chn[CHNrdr] = UR2in;
	break;
    }
    switch (io & 3) {
    case 0: chn[CHNconin] = TTYin;
	chn[CHNconout] = TTYout;
	break;
    case 1: chn[CHNconin] = CRTin;
	chn[CHNconout] = CRTout;
	break;
    case 2: chn[CHNconin] = chn[CHNrdr];
	chn[CHNconout] = chn[CHNlst];
	break;
    case 3: chn[CHNconin] = UC1in;
	chn[CHNconout] = UC1out;
	break;
    }
    old_iobyte = io;
}

static struct sio *
getsiop(int chan)
{
    if (ram[3] != old_iobyte)
	new_iobyte(ram[3]);
    return &siotab[chn[chan]];
}

/* gr: */
int isafifo(int desc)
{
    struct stat buf;
    
    return fstat(desc, &buf) == 0 && S_ISFIFO(buf.st_mode);
}

int
bios_init(const char *initfile)
{
    int go = 0;
    FILE *su;
    const char *fn = initfile;
    char *name;
    char buf[BUFSIZ];

#ifdef DEBUG
    signal(SIGINT, sighand);
#endif
    if (z3env && z3env > bios_top && z3env < dptable) {
	/* reserve space for a z-system environment page */
	WORD p = z3env-bios_top;
	int i = 1024;
	while (i--) {
	    global_alv[p>>3] |= (0x80 >> (p&7));
	    p++;
	}
    }
    siotab[CRTin].fp = stdin;
    name = ttyname(fileno(stdin));
    siotab[CRTin].filename = name ? newstr(name) : "(stdin)";
    siotab[CRTin].tty = isatty(fileno(stdin));
    /* gr: */
    siotab[CRTin].canselect = siotab[CRTin].tty || isafifo(fileno(stdin));
    siotab[CRTout].fp = stdout;
    name = ttyname(fileno(stdout));
    siotab[CRTout].filename = name ? newstr(name) : "(stdout)";
    siotab[CRTout].tty = isatty(fileno(stdout));
    /* gr: */
    siotab[CRTout].canselect = siotab[CRTout].tty || isafifo(fileno(stdout));
    if (siotab[CRTin].tty) {
	ttyflags = ISATTY;
	if (tcgetattr(fileno(stdin), &cookedtio) != 0) {
	    perror("tcgetattr");
	    exit(1);
	}
	rawtio = cookedtio;
	rawtio.c_iflag = 0;
	rawtio.c_oflag = 0;
	rawtio.c_lflag = interrupt ? ISIG : 0;
	memclr(rawtio.c_cc, NCCS);
	rawtio.c_cc[VINTR] = interrupt;
	rawtio.c_cc[VMIN] = 1;
    }
    if (*initfile != '/' && access(initfile, R_OK) != 0) {
	char *h = getenv("HOME");
	if (h) {
	    fn = buf;
	    sprintf((char *) fn, "%s/%s", h, initfile);
	}
    }
    if ((su = fopen(fn, "r")) == NULL)
	return 0;
    while (!go && fgets(buf, BUFSIZ-1, su))
	go = docmd(buf);
    fclose(su);
    atexit(ttycook);
    ttyraw();
    return go;
}

/* The constat() routine is complicated by the need to deal with CP/M programs
   that poll console status without doing any other bios calls.
   Examples are Wordstar and Backgrounder II.  In Wordstar the purpose is to
   act as a crude timer that can be interrupted by the user pressing a key
   (delay before a menu is shown).  In Backgrounder it's an artifact of the
   pseudo multitasking logic.

   A simple-minded constat() implementation leads to such CP/M programs soaking
   up 100% of the host CPU and getting confused about the system speed.

   The 2 constants CSTTIME and CSTCOUNT control the emulation.
   constat() normally returns the current console status immediately.
   However, after it has been called CSTCOUNT times without any other bios
   activity, it waits CSTTIME microseconds (or until a character is available)
   before returning.
*/

#define CSTTIME		100000
#define CSTCOUNT	128

static const struct timeval immediate = { 0, 0 };
static const struct timeval delay = { 0, CSTTIME };

static int
constat(void)
{
    static int consecutive;		/* number of consecutive calls */
    static COUNT lastcount;		/* previous call */
    struct timeval t = immediate;
    fd_set rdy;
    struct sio *s = getsiop(CHNconin);
    int fd;

    if (s->fp == NULL)		/* no file */
	return 0;
    if (s->canselect == 0)	/* disk files are not always ready! (gr) */
	return 1;
    if (bioscount != lastcount+1)
	consecutive = 0;
    else if (++consecutive == CSTCOUNT) {
	consecutive = 0;
	t = delay;
    }
    lastcount = bioscount;

    fd = fileno(s->fp);
    FD_ZERO(&rdy);
    FD_SET(fd, &rdy);
    (void) select(fd+1, &rdy, NULL, NULL, &t);
    return FD_ISSET(fd, &rdy);
}

static int
lststat(void)
{
    static int consecutive;		/* number of consecutive calls */
    static COUNT lastcount;		/* previous call */
    struct timeval t = immediate;
    fd_set rdy;
    struct sio *s = getsiop(CHNlst);
    int fd;

    if (s->fp == NULL)			/* no file */
	return 0;
    if (s->tty == 0)			/* disk files are always ready */
	return 1;
    if (bioscount != lastcount+1)
	consecutive = 0;
    else if (++consecutive == CSTCOUNT) {
	consecutive = 0;
	t = delay;
    }
    lastcount = bioscount;

    fd = fileno(s->fp);
    FD_ZERO(&rdy);
    FD_SET(fd, &rdy);
    (void) select(fd+1, &rdy, NULL, NULL, &t);
    return FD_ISSET(fd, &rdy);
}

static int
serin(int chan)
{
    char c;
    int ch;
    struct sio *s = getsiop(chan);

    if (s->fp == NULL)
	return 0x1a;
    if (s->tty) {
	if (read(fileno(s->fp), &c, 1) == 0)
	    return 0x1a;
	else
	    return c;
    }
    if ((ch = getc(s->fp)) == EOF)
	return 0x1a;
    else
	return ch & 0xff;
}

static void
serout(int chan, char c)
{
    struct sio *s = getsiop(chan);

    if (s->fp == NULL)
	return;
    if (s->tty)
	(void) write(fileno(s->fp), &c, 1);
    else
	fwrite(&c, 1, 1, s->fp);
}

static WORD
seldsk(int disk)
{
    if (disk < 0 || disk > 15 || !(mnttab[disk].flags & MNT_ACTIVE))
	return 0;
    curdisk = mnttab + disk;
    return curdisk->dph;
}

#ifdef BGii_BUG
/* I cannot persuade Backgrounder ii to output the H at the end of a
   VT-100/xterm cursor-positioning sequence.  This kludges it. */
void
BGiiFix(int c)
{
    static const char cc[] = "\033[0;0";
    static const char *p = cc;

    if (c == 0)
	return;
    if ('0' <= c && c <= '9')
	return;
    if (*p == '0')
	p++;
    if (*p == c) {
	p++;
	return;
    }
    if (*p) {
	p = cc;
	return;
    }
    p = cc;
    if (c == 'H')
	return;
    serout(CHNconout, 'H');
}
#else
#define BGiiFix(x)
#endif

#define FCACHE_SIZE	4
static struct fc {
    FILE *f;
    struct mnt *disk;
    struct fdesc *fd;
} fcache[FCACHE_SIZE];

/* clear the file cache when disk is unmounted */
void
clearfc(struct mnt *dp)
{
    int i;

    for (i = 0; i < FCACHE_SIZE; i++) {
	struct fc *f = fcache+i;
	if (f->f && f->disk == dp) {
	    fclose(f->f);
	    f->f = NULL;
	}
    }
}

static int
readsec(void)
{
    unsigned long offset, blockno;

    if (curdisk == NULL || !(curdisk->flags & MNT_ACTIVE))
	return 1;
    offset = (track*GetWORD(curdisk->dph+16) + sector)*128;
    blockno = offset / BLOCK_SIZE;
    if ((curdisk->flags & MNT_UNIXDIR) == 0) {
	/* the 'disk image' case is easy */
	memcpy(ram + dma, curdisk->data + offset, 128);
	return 0;
    }

    /* handle a 'unix directory' disc */
    if (blockno < 16) {
	/* directory access */
	if (offset/32 < curdisk->nde)
	    memcpy(ram+dma, curdisk->data+offset, 128);
	else
	    memset(ram+dma, 0xe5, 128);
    }
    else {			/* file access */
	int i;
	for (i = 0; i < FCACHE_SIZE; i++) {
	    struct fc *f = fcache + i;
	    if ((f->f == NULL) ||
		(f->disk == curdisk && blockno >= f->fd->firstblock &&
		 blockno <= f->fd->lastblock))
		break;
	}
	if (i == FCACHE_SIZE) {
	    /* cache overflow */
	    fclose(fcache[FCACHE_SIZE-1].f);
	    memmove(fcache+1, fcache, (FCACHE_SIZE - 1) * sizeof(struct fc));
	    i = 0;
	    fcache[0].f = NULL;
	}
	if (fcache[i].f == NULL) {
	    struct fc *f = fcache + i;
	    struct fdesc *fd = curdisk->fds;
	    int j;
	    f->disk = curdisk;
	    for (j = 0; j < curdisk->nfds; j++, fd++)
		if (blockno >= fd->firstblock && blockno <= fd->lastblock)
		    break;
	    if (j >= curdisk->nfds) {
		/* cant happen */
		return 1;
	    }
	    if ((f->f = fopen(fd->fullname, "rb")) == NULL) {
		perror(fd->fullname);
		return 1;
	    }
	    f->fd = fd;
	}
	if (i != 0) {
	    /* sort least-recently-used to front */
	    struct fc temp;
	    temp = fcache[i];
	    memmove(fcache+1, fcache, i * sizeof(struct fc));
	    fcache[0] = temp;
	}
	if (fseek(fcache[0].f, offset-fcache[0].fd->firstblock*BLOCK_SIZE,
		  SEEK_SET) < 0)
	    return 1;
	if ((i = fread(ram+dma, 1, 128, fcache[0].f)) < 128)
	    memset(ram+dma+i, 0x1a, 128-i); /* EOF */
    }
    return 0;
}


static int
writesec(int stype)
{
    if (curdisk == NULL ||
	((curdisk->flags & (MNT_ACTIVE|MNT_RDONLY)) != MNT_ACTIVE))
	return 1;
    memcpy(curdisk->data +
	   (track*GetWORD(curdisk->dph+16) + sector)*128, ram + dma, 128);
    return 0;
}

/* This only works for single-byte sector numbers
   (but then, so does the CP/M 2.2 CBIOS code).
*/
static WORD
sectrans(WORD sec, WORD table)
{
    if (table == 0)
	return sec;
    return GetBYTE(table + sec);
}

static BYTE savcpm[CPM_LENGTH];		/* saved copy of cpm */

void
bios(int func)
{
    bioscount++;
    switch (func) {
    case 0:			/* cold boot */
	memcpy(savcpm, ram + ccp_base, CPM_LENGTH);
	goto gocpm;
    case 1:			/* warm boot */
	memcpy(ram + ccp_base, savcpm, CPM_LENGTH);
    gocpm:
	dma = 0x80;
	ram[0] = 0xc3;
	ram[1] = 0x03;
	ram[2] = hreg(bios_base);
	ram[5] = 0xc3;
	ram[6] = 0x06;
	ram[7] = hreg(bdos_base);
	Setlreg(BC, ram[4]);
	pc = ccp_base;
	return;
    case 2:			/* console status */
	Sethreg(AF, constat() ? 0xff : 0x00);
	break;
    case 3:			/* console input */
	Sethreg(AF, serin(CHNconin) & 0x7f);
	break;
    case 4:			/* console output */
	BGiiFix(lreg(BC));
	serout(CHNconout, lreg(BC));
	break;
    case 5:			/* list output */
	serout(CHNlst, lreg(BC));
	break;
    case 6:			/* punch output */
	serout(CHNpun, lreg(BC));
	break;
    case 7:			/* tape reader input */
	Sethreg(AF, serin(CHNrdr));
	break;
    case 8:			/* home disk */
	track = 0;
	break;
    case 9:			/* select disk */
	HL = seldsk(lreg(BC));
	break;
    case 10:			/* set track */
	track = BC;
	break;
    case 11:			/* set sector */
	sector = BC;
	break;
    case 12:			/* set dma */
	dma = BC;
	break;
    case 13:			/* read sector */
	Sethreg(AF, readsec());
	break;
    case 14:			/* write sector */
	Sethreg(AF, writesec(lreg(BC)));
	break;
    case 15:			/* list status */
	Sethreg(AF, lststat() ? 0xff : 0x00);
	break;
    case 16:			/* translate sector */
	HL = sectrans(BC, DE);
	break;
    case 253:			/* return from recursive call */
	return;
    case 254:
	++pc;			/* meta-level command */
	if (GetBYTE(pc) == 0)
	    monitor(0);
	else {
	    /* need a copy because docmd() scratches its argument */
	    char *sav = newstr((char *) ram+pc);
	    ttycook();
	    (void) docmd(sav);
	    ttyraw();
	    free(sav);
	    pc += strlen((char *) ram+pc);
	}
	break;
    case 255:			/* quit */
	exit(0);
	break;
    default:
	fprintf(stderr, "Invalid bios function: %d\r\n", func);
    }
    pc++;
}
